### PR TITLE
Day 14 - Add Year-Month Validation, Adjust Visualization for Edge Cases

### PR DESCRIPTION
**Pull request by kishanpatel789**
## Change Log
- Add validation of `year_month` to API reports route. Used recommendation from [FastAPI docs](https://fastapi.tiangolo.com/tutorial/query-params-str-validations/#custom-validation).
- Add tests for `year_month` validation.
- Update data visualization for edge cases (eg category has no budget, etc)
- Suppress "unresolved attribute" error in `ty` for SQLModel related issues only.
- Improve layout of Transaction forms in NiceGUI.

## TODO 
- [ ] Begin dockerization of services (Postgres, FastAPI, NiceGUI)
- [ ] Write documentation for project

### CODE REVIEW

File: src/api/helpers.py [hunk 1]
```diff
@@ -135,6 +143,16 @@ def create_page(
     return page_output
 
 
+def validate_year_month(year_month: str) -> str:
+    year, month = map(int, year_month.split("-"))
+    if year > date.today().year:
+        raise ValueError("Input year cannot exceed current year.")
+    if not (1 <= month <= 12):
```
Comment (bbelderbos) [inline src/api/helpers.py:150]: another way is to try to construct a date object, catching and re-raising the exception:

```
try:
    year, month = map(int, year_month.split("-"))
    date(year, month, 1)
except ValueError:
    raise ValueError("Invalid year-month format or values. Use 'YYYY-MM'.")
```

I repeat ValueError here to add my own error message

File: src/api/helpers.py [hunk 2]
```diff
@@ -12,8 +12,16 @@
 
 from .models import PageBase, PageLinks, PaginationInput
 
-ProjectInfo = namedtuple("ProjectInfo", ["version", "author_name", "author_email"])
-DateRange = namedtuple("DateRange", ["start", "end"])
+
+class ProjectInfo(NamedTuple):
```
Comment (bbelderbos) [inline src/api/helpers.py:16]: nice

File: src/api/routers/reports.py [hunk 3]
```diff
@@ -16,11 +17,13 @@
     tags=["reports"],
 )
 
+YearMonthParam = Annotated[
+    str, Query(pattern=r"\d{4}-\d{2}"), AfterValidator(validate_year_month)
```
Comment (bbelderbos) [inline src/api/routers/reports.py:21]: as discussed, gradual validation with early more generic bailing, nice

File: src/api/routers/transactions.py [hunk 4]
```diff
@@ -60,9 +60,9 @@ def read_transactions(
         search_term = f"%{q}%"
         query = query.where(
             or_(
-                Transaction.vendor.ilike(search_term),
-                Transaction.note.ilike(search_term),
-                Transaction.category.has(Category.name.ilike(search_term)),
+                Transaction.vendor.ilike(search_term),  # ty: ignore[unresolved-attribute]
+                Transaction.note.ilike(search_term),  # ty: ignore[unresolved-attribute]
+                Transaction.category.has(Category.name.ilike(search_term)),  # ty: ignore[unresolved-attribute]
```
Comment (bbelderbos) [inline src/api/routers/transactions.py:65]: tried on or_ but would not work
will check https://docs.astral.sh/ty/suppression/#ty-suppression-comments, but seems you could not find more general approach without being too permissive then

File: src/gui/pages/home.py [hunk 5]
```diff
@@ -14,13 +14,26 @@ def home():
...
                 bar_color = mint_green if is_under else mint_red
 
-                value_text = f"${value:,} of ${budget:,}"
+                if budget_str is not None:
+                    delta_amount = round(abs(budget - value))
+                    delta_text = f"${delta_amount:,} {'left' if is_under else 'over'}"
+                else:
+                    delta_text = ""
+
+                if title is None:
```
Comment (bbelderbos) [inline src/gui/pages/home.py:35]: would it ever be None? the type hint says `str`

File: src/gui/pages/transactions.py [hunk 6]
```diff
@@ -145,17 +145,17 @@ def open_create_modal():
...
                         # category drop-down selector
                         category_id = ui.select(
                             label="Category",
                             options=get_selectable_categories(),
                             with_input=True,
-                        ).classes("w-full")
+                        ).classes("w-4/10")
                     with ui.row():
-                        note = ui.textarea("Note").classes("w-full")
+                        note = ui.textarea("Note").classes("w-10/10")
```
Comment (bbelderbos) [inline src/gui/pages/transactions.py:158]: I will check this ...

**Review by bbelderbos — APPROVED** [review]
Nice work.

File: src/api/routers/transactions.py (hunk 4 above)
Comment (kishanpatel789) [reply src/api/routers/transactions.py:65]: Yes, from some basic experiments, it seems like the suppression comments (`# ty: ignore[...]`) only operate line by line, not for a section of code.

File: src/gui/pages/transactions.py (hunk 6 above)
Comment (bbelderbos) [reply src/gui/pages/transactions.py:158]: ok, can you check this out? https://github.com/kishanpatel789/finance_tracker/pull/35

Comment (kishanpatel789) [reply src/gui/pages/transactions.py:158]: This is fantastic! Thanks Bob!
